File: Unity/Assets/Control/PID.cs


public class PID
{
    private float Kp;
    private float Kd;
    private float Ki;
    private float error = 0;
    private float prevError = 0;
    public float integ = 0;
    private float  deriv = 0;
    private float desired = 0;
    public float outP;       
    public float outI;      
    public float outD;        
    private float iLimit;   

    public PID(float Kp,float Kd,float Ki)
    {
        this.Kp = Kp;
        this.Kd = Kd;
        this.Ki = Ki;
    }

    public float PIDupdate(float measured, float expect)
    {
        float output;

        this.desired = expect;                        

        this.error = this.desired - measured;         

        this.integ += this.error;

        if (this.integ > this.iLimit)          
        {
            this.integ = this.iLimit;
        }
        else if (this.integ < -this.iLimit)    
        {
            this.integ = -this.iLimit;
        }
        this.deriv = this.error - this.prevError;     

        this.outP = this.Kp * this.error;           
        this.outI = this.Ki * this.integ;
        this.outD = this.Kd * this.deriv;

        output = this.outP +
                         this.outI +
                         this.outD;

        this.prevError = this.error;                               

        return output;
    }

    public void pidSetIntegralLimit(float limit)
    {
        this.iLimit = limit;
    }
}

File: Flight_Controller/Q-Flight_F4/Control/PID.c
#include "PID.h"


/*------------------------------------------pid结构初始化-----------------------------------------*/
//输入参数：结构体指针，期望值，kp,ki,kd
void pidInit(pidsuite* pid, const float desired, const float kp,
             const float ki, const float kd)
{

  pid->error = 0;
  pid->prevError = 0;
  pid->integ = 0;
  pid->deriv = 0;
  pid->desired = desired;
  pid->kp = kp;
  pid->ki = ki;
  pid->kd = kd;
  
  pid->iLimit = DEFAULT_PID_INTEGRATION_LIMIT;
}

/*----------------------------------------------pid积分部分限制值-------------------------------------------*/
void pidSetIntegralLimit(pidsuite* pid, const float limit)
{
  pid->iLimit = limit;
}

float pidErrorUpdate(pidsuite* pid, const float expect,float error) // target - current
{
  float output;
	static uint8_t integ_sign=0;
	uint8_t is_shrink=1;

  pid->desired=expect;			 				//获取期望角度

  pid->error = error;	 	  //偏差：期望-测量值
	if(is_shrink){
		if(integ_sign==0){
			pid->integ += pid->error;
			integ_sign=1;
		}
		else{
			if(math_sign(pid->error)!=math_sign(pid->integ)){
				pid->integ += pid->error;	  //偏差积分
			}
		}
	}
	else{
		pid->integ += pid->error;	
	}
  if (pid->integ > pid->iLimit)				  //积分限幅
  {
    pid->integ = pid->iLimit;
  }
  else if (pid->integ < -pid->iLimit)		//积分限幅
  {
    pid->integ = -pid->iLimit;
  }
  pid->deriv = pid->error - pid->prevError;		//微分

	pid->outP = pid->kp * pid->error;					  //方便独立观察
	pid->outI = pid->ki * pid->integ;
	pid->outD = pid->kd * pid->deriv;
	
	output = pid->outP +
					 pid->outI +
					 pid->outD;
  
  pid->prevError = pid->error;							 		//更新前一次偏差

  return output;
}

/*--------------------------------------pid输出更新------------------------------------------*/
//输入参数：pid结构体指针，测量值 ,期望值
//输出：pid输出
float pidUpdate(pidsuite* pid, const float measured,float expect)
{
  float output;
	static uint8_t integ_sign=0;
	uint8_t is_shrink=1;

  pid->desired=expect;			 				//获取期望角度

  pid->error = pid->desired - measured;	 	  //偏差：期望-测量值
	if(is_shrink){
		if(integ_sign==0){
			pid->integ += pid->error;
			integ_sign=1;
		}
		else{
			if(math_sign(pid->error)!=math_sign(pid->integ)){
				pid->integ += pid->error;	  //偏差积分
			}
		}
	}
	else{
		pid->integ += pid->error;	
	}
  if (pid->integ > pid->iLimit)				  //积分限幅
  {
    pid->integ = pid->iLimit;
  }
  else if (pid->integ < -pid->iLimit)		//积分限幅
  {
    pid->integ = -pid->iLimit;
  }
  pid->deriv = pid->error - pid->prevError;		//微分

	pid->outP = pid->kp * pid->error;					  //方便独立观察
	pid->outI = pid->ki * pid->integ;
	pid->outD = pid->kd * pid->deriv;
	
	output = pid->outP +
					 pid->outI +
					 pid->outD;
  
  pid->prevError = pid->error;							 		//更新前一次偏差

  return output;
}

